[RUM][REPLAY] update content for "privacy by default" feature. (#11677)
DataDog/documentation: content/en/real_user_monitoring/guide/session-replay-getting-started.md

```javascript
if (user.isAuthenticated) {
    DD_RUM.startSessionReplayRecording();
}
```

The Session Replay recording can be stopped by calling `stopSessionReplayRecording()`.

### Sensitive & personal data obfuscation

<div class="alert alert-info">
To be released < href="https://github.com/DataDog/browser-sdk/pull/1049">at version 3.5.0 on September 15, 2021<a/>, the new privacy-by-default functionality will change how the SDK manages privacy.
</div>

In short, this means that the SDK protects end-user privacy and sensitive organisation information from being recorded by default, automatically masking form fields like password inputs and text areas.

You may configure the default privacy mode within your Javascript configuration and tag specific parts of your HTML documents with explicit overrides. There are three main points to understand:

1. There are four privacy levels supported: `allow`, `mask-user-input`, `mask`, and `hidden`

> 2. The root HTML element inherits the privacy level from the JS configuration `defaultPrivacyLevel` property which defaults to `mask-user-input` (for example `{defaultPrivacyLevel: 'allow'}`)

Reviewer: I'd use the default value as an example:

```suggestion
2. The root HTML element inherits the privacy level from the JS configuration `defaultPrivacyLevel` property which defaults to `mask-user-input` (e.g: `{defaultPrivacyLevel: 'mask-user-input'}`)
```

3. Each HTML element inherits the privacy level of its parent (except `hidden` which is final)

Understanding the inheritance rules above, you may tag the privacy level of an HTML element one of two ways:

1. using an HTML attribute (such as `data-dd-privacy="allow"` or `data-dd-privacy="mask"`); or
2. using an HTML class name (such as `class="dd-privacy-allow"` or `class="dd-privacy-mask-user-input"`).

> Now let's unpack how each of the four privacy levels behave:

Reviewer: ```suggestion
The behavior of each of the four privacy levels is as follows:
```

> -   `allow`: generally unmasks everything, excluding HTML input elements of type `password`, `email`, and `tel`, and elements with `autocomplete` attributes.

Reviewer: ```suggestion
-   `allow`: generally records everything unmasked, with the exception of HTML input elements of type `password`, `email`, and `tel`, and elements with `autocomplete` attributes.
```

> -   `mask-user-input`: unmasks general HTML content while blocking most form fields like inputs, textareas, checkbox values etc.

Reviewer: ```suggestion
-   `mask-user-input`: masks most form fields like inputs, textareas, checkbox values, etc. while recording all other text as is.
```

> -   `mask`: masks all HTML text, form values, images, and links.

Reviewer: ```suggestion
-   `mask`: masks all HTML text, user input, images, and links.
```

> -   `hidden`: blocks session replay nearly entirely- suppressing the recording of all JS events, records only the dimension of the element, and all child elements will be untracked- and cannot override this `hidden` privacy level.

Reviewer: ```suggestion
-   `hidden`: skips recording the element entirely, and this mode cannot be overridden in a child node.
```

When tagging your web application, the safest strategy is to start by tagging `mask` at the top of your HTML document and allowing each team to consider which pages, features, and components they need unmasked (`allow`) to accomplish their work.

#### Example

> Any input values within the following form are be replaced with asterisks "\*\*\*\* \*\* \*\*\*".

Reviewer 2: ```suggestion
Input values in the following form are replaced with asterisks `\*\*\*\* \*\* \*\*\*`.
```

```html
<form method="post" data-dd-privacy="mask-user-input">
    <input type="text" name="name" id="name" required />
    <input type="number" name="age" id="age" required />
    <input type="email" name="email" id="email" required />
    <input type="submit" value="submit" />
</form>
```

> Alternatively, HTML elements can be fully obfuscated with `hidden`:

Reviewer: ```suggestion
Alternatively, HTML elements can be fully skipped from recording with `hidden`:
```

```html
<div id="profile-info" data-dd-privacy="hidden">
    <p>Name: John Doe</p>
    <p>Birth date: June 6th, 1987</p>
</div>
```

> These elements are replaced with a gray block at recording time, as seen in this example of the navbar of a replay of a Datadog session:

Reviewer 2: ```suggestion
These elements are replaced with a gray block at recording time. In this example replay session, the navigation in Datadog is obfuscated:

```

You can configure the default privacy mode within your Javascript configuration and tag specific parts of your HTML documents with explicit overrides.

- Four privacy levels are supported: `allow`, `mask-user-input`, `mask`, and `hidden`.
- The root HTML element inherits the privacy level from the JavaScript configuration `defaultPrivacyLevel` property which defaults to `mask-user-input`, such as `{defaultPrivacyLevel: 'mask-user-input'}`.
- Every HTML element inherits the privacy level of its parent except for `hidden`, which is final.

With the inheritance rules above, you can tag the privacy level of an HTML element using one of two methods:

1. An HTML attribute such as `data-dd-privacy="allow" | "mask" | "hidden" | "mask-user-input"`; or
2. An HTML class name such as `class="dd-privacy-allow" | "dd-privacy-mask-user-input" | "dd-privacy-mask" | "dd-privacy-hidden"`.

Every privacy level behaves as:

-   `allow`: Unmasks everything, excluding HTML input elements such as `password`, `email`, and `tel`, and elements with `autocomplete` attributes.

> -   `mask-user-input`: Unmasks HTML content, blocking most form fields such as inputs, text areas, and checkbox values.

Reviewer 3: ```suggestion
-   `mask-user-input`: Unmasks HTML content, but blocks most form fields such as inputs, text areas, and checkbox values. Inputs are replaced with three asterisks (`***`) and text areas are obfuscated with space-preserving `x` characters.
```

> -   `allow`: Unmasks everything, excluding HTML input elements such as `password`, `email`, and `tel`, and elements with `autocomplete` attributes.
> -   `mask-user-input`: Unmasks HTML content, but blocks most form fields such as inputs, text areas, and checkbox values. Inputs are replaced with three asterisks (`***`) and text areas are obfuscated with space-preserving `x` characters.
> -   `mask`: Masks all HTML text, form values, images, and links.
> -   `hidden`: Blocks Session Replay. Blocking Session Replay suppresses the recording of all JavaScript events and records only the dimensions of the element. All child elements are untracked. By default, `hidden` cannot be overridden by a child element.

Reviewer 2: ```suggestion
-   `allow`: Records everything unmasked, excluding HTML input elements such as `password`, `email`, and `tel`, and elements with `autocomplete` attributes.
-   `mask-user-input`: Masks most form fields such as inputs, text areas, and checkbox values while recording all other text as is. Inputs are replaced with three asterisks (`***`) and text areas are obfuscated with space-preserving `x` characters.
-   `mask`: Masks all HTML text, user input, images, and links.
-   `hidden`: Skips recording the element completely. By default, `hidden` cannot be overridden in a child node.
```

When tagging your web application, start by tagging `mask` at the top of your HTML document and ask your team to consider which pages, features, and components need to be unmasked (`allow`).

#### Example

> Input values in the following form are replaced with asterisks, for example: `\*\*\*\* \*\* \*\*\*`.

Reviewer 3: ```suggestion
Input values in the following form are replaced with three asterisks (`***`).
```

1. The resource does not exist anymore. For example, it was part of a previous deployment.
2. The resource is inaccessible. For example, authentication might be required, or the resource might only be accessible from an internal network.

> 3. The resource is blocked by the browser due to CORS (typically, web-fonts).

Reviewer 2: ```suggestion
3. The resource is blocked by the browser due to CORS (typically web-fonts).
```

   The replay being rendered on the `app.datadoghq.com` domain, the assets requests are subject to cross origin security checks by your browser. If the given asset is not authorised for the domain, your browser blocks the request.
   The fix is thus to allow `app.datadoghq.com` through the [`Access-Control-Allow-Origin`][5] header for any font or image assets your website depends upon, ensuring these resources are accessible for the replay.
   To learn more about Cross Origin Resource Sharing, see the [MDN Web Docs article][6].

> ### CSS rules not properly applied / mouse hover not replayed

Reviewer 2: ```suggestion
### CSS rules not properly applied/mouse hover not replayed
```

### What is the performance impact?

> The team is is focused on ensuring minimal impact of the Session Replay recorder on the applications performance:

Reviewer 2: ```suggestion
To ensure minimal impact of the Session Replay recorder on the application's performance:
```

> -   reduced network impact of Session Replay by compressing the data prior to sending it to Datadog
> -   reduced load on the browser’s UI thread by delegating most of the CPU intensive work (like compression) to a background service worker.

Reviewer 2: ```suggestion
-   Reduce network impact of Session Replay by compressing the data prior to sending it to Datadog.
-   Reduce load on the browser’s UI thread by delegating most of the CPU intensive work (such as compression) to a background service worker.
```
